File: ruby/day11.rb
require "set"

def main
  octopi = $stdin.readlines.map { |line| line.strip.chars.map(&:to_i) }
  flashes = 0
      #puts "Before any steps:"
      #puts octopi.map(&:join)
  100.times do |i|
    octopi, inc = step(octopi)
    flashes += inc
    if i < 10
      #puts "After step #{i+1}:"
      #puts octopi.map(&:join)
    end
  end
  puts "part 1: #{flashes}"

  i = 100
  loop do
    i += 1
    octopi, inc = step(octopi)
    if i > 190 && i < 200
      puts "After step #{i}:"
      puts octopi.map(&:join)
    end
    if inc == 100
      puts "part 2: #{i}"
      break
    end
  end
end

def step(octopi)
  flashes = []
  (0..9).each do |i|
    (0..9).each do |j|
      if (octopi[i][j] += 1) == 10
        flashes << [i,j]
      end
    end
  end
  flashed = []
  while c = flashes.shift
    #p flash: c
    neighbors(*c).each do |i, j|
      if (octopi[i][j] += 1) == 10
        flashes << [i, j]
      end
    end
    flashed.push c
  end
  #puts "flashed:"
  #puts flashed.sort.map { |c| " - #{c.inspect}" }
  flashed.each do |i,j|
    octopi[i][j] = 0
  end
  [octopi, flashed.size]
end

def neighbors(x,y)
  [
    [x+1, y],
    [x-1, y],
    [x, y+1],
    [x, y-1],
    [x+1, y+1],
    [x-1, y+1],
    [x+1, y-1],
    [x-1, y-1],
  ].select { |i,j| j < 10 && i < 10 && i >= 0 && j >= 0 }
end

main

File: src/day11.rs
use super::common::read_lines::read_lines;
use std::io::Read;

pub fn run<R: Read>(r: R) {
    let mut octopi = [[0; 10]; 10];
    for (i, line) in read_lines(r).enumerate() {
        octopi[i] = parse_line(&line).unwrap();
    }

    let mut steps = 0;
    let mut flashes = 0;
    loop {
        let (next, flashed) = step(octopi);
        steps += 1;
        flashes += flashed;
        octopi = next;
        if steps == 100 {
            println!("part 1: {}", flashes);
        }
        if flashed == 100 {
            println!("part 2: {}", steps);
            return;
        }
    }
}

fn parse_line(line: &str) -> Result<[u8; 10], String> {
    let mut row = [0; 10];
    for (j, c) in line.chars().enumerate() {
        if j > 9 {
            return Err(format!("too many chars ({}) in {}", c, line));
        }
        row[j] = (c as u8) - b'0';
    }
    Ok(row)
}

fn step(mut octopi: [[u8; 10]; 10]) -> ([[u8; 10]; 10], usize) {
    let mut flashed = vec![];
    for (i, row) in octopi.iter_mut().enumerate() {
        for (j, val) in row.iter_mut().enumerate() {
            *val += 1;
            if *val == 10 {
                flashed.push((i, j));
            }
        }
    }
    let mut x = 0;
    while x < flashed.len() {
        for (i, j) in neighbors(flashed[x]) {
            octopi[i][j] += 1;
            if octopi[i][j] == 10 {
                flashed.push((i, j));
            }
        }
        x += 1;
    }
    for (i, j) in flashed.iter() {
        octopi[*i][*j] = 0;
    }
    (octopi, flashed.len())
}

fn neighbors(coords: (usize, usize)) -> Neighbors {
    Neighbors {
        coords: (coords.0 as isize, coords.1 as isize),
        di: -1,
        dj: -1,
    }
}

struct Neighbors {
    coords: (isize, isize),
    di: isize,
    dj: isize,
}

impl Iterator for Neighbors {
    type Item = (usize, usize);

    fn next(&mut self) -> Option<Self::Item> {
        if self.di > 1 || self.dj > 1 {
            None
        } else {
            let (ci, cj) = self.coords;
            let (i, j) = ((ci + self.di), (cj + self.dj));
            if self.dj == 1 {
                self.di += 1;
                self.dj = -1;
            } else {
                self.dj += 1;
            }
            if (i == ci && j == cj) || i < 0 || i > 9 || j < 0 || j > 9 {
                self.next()
            } else {
                Some((i as usize, j as usize))
            }
        }
    }
}
